zeebe/engine/src/main/java/io/camunda/zeebe/engine/state/authorization/DbAuthorizationState.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Camunda License 1.0. You may not use this file
+ * except in compliance with the Camunda License 1.0.
+ */
+package io.camunda.zeebe.engine.state.authorization;
+
+import io.camunda.zeebe.db.ColumnFamily;
+import io.camunda.zeebe.db.TransactionContext;
+import io.camunda.zeebe.db.ZeebeDb;
+import io.camunda.zeebe.db.impl.DbCompositeKey;
+import io.camunda.zeebe.db.impl.DbLong;
+import io.camunda.zeebe.db.impl.DbString;
+import io.camunda.zeebe.engine.state.immutable.AuthorizationState;
+import io.camunda.zeebe.engine.state.mutable.MutableAuthorizationState;
+import io.camunda.zeebe.protocol.ZbColumnFamilies;
+import io.camunda.zeebe.protocol.impl.record.value.authorization.AuthorizationRecord;
+
+public class DbAuthorizationState implements AuthorizationState, MutableAuthorizationState {
+  private final PersistedAuthorization persistedAuthorization = new PersistedAuthorization();
+
+  private final PersistedPermissions persistedPermissions = new PersistedPermissions();
+
+  private final DbLong authorizationKey;
+  private final DbString ownerKey;
+  private final DbString ownerType;
+  private final DbString resourceKey;
+  private final DbString resourceType;
+  private final DbCompositeKey<DbString, DbString> resourceCompositeKey;
+  private final DbCompositeKey<
+          DbCompositeKey<DbString, DbString>, DbCompositeKey<DbString, DbString>>
+      ownerAndResourceCompositeKey;
+  private final ColumnFamily<DbLong, PersistedAuthorization> authorizationColumnFamily;
+  private final ColumnFamily<
+          DbCompositeKey<DbCompositeKey<DbString, DbString>, DbCompositeKey<DbString, DbString>>,
+          PersistedPermissions>
```

Comment:
Great question @remcowesterhoud, If we take the owner type as an example, the values could be any one of 
``` USER, ROLE, GROUP```
what I cannot say is if there would be a `role` with the same key as a `group`. For now I'd opt for keeping this complex key to be certain on the uniqueness and use a comment but if you feel strongly otherwise I can change it or follow up after

Reply:
Regarding resource key, I'm not 100% sure, maybe thats something you could provide input on, resource type here is any resource in the system i.e. `authorization`, `process definition` and so on.

Reply:
We can indeed, so roles and groups are created in the UI in SM, in SaaS the roles are taken from Console and Groups are created via an Identity component within Console. They have a very minimal set of fields Name, Description.

Currently the unique key of these are the name however they are unique _within_ their context (due to the table split. When an Authorization is created currently it is unique in our databases with the following fields 
```
"entityId", "entityType", "organizationId", "resource_id", "permission"
```
So here the only new field resource type (entityId/entityType maps to ownerKey/ownerType).

For this reason I believe that we need the distinction of ownerType because its very plausible that a group and role could cross over in name IMO.

Resource type is something that I'd need your advice on for the wider platform picture, we need to store the type at least so we can perform targeted `get()` calls for a specific resource type when checking authorizations at a command level however if the type needs to be in the key is unclear to me. The resourceType here can be though of _any_ resource we have in the platform so `Authorization`, `Group`, `Process Definition`, etc... so I'm not sure I can confidently say that there will never be an overlap in resource key especially if we use the name of the entity as a key and not have Zeebe generate the key.

Reply:
So, I had a good think about how we should store Authorizations in a nice way that makes it easy and hopefully performant to verify user permissions later on. Bear with me here 😄 

**ColumnFamily**
I believe we should have a ColumnFamily that allows easy searching by ownerKey. My proposal is:
`ownerKey - resourceType - permissionType -> Authorizations`

An example key would be `12345ProcessInstanceCREATE`. With this key we can do a targeted search for the authorizations of this owner, for this resource and with this permissionType. An `ownerKey` will be the key of a user, a group or a role.

The Authorizations object that's the value will contain a list of resource keys (eg: `bpmnProcessId:*`). Potentially we could even split this is another object where we store the `bpmnProcessId` and the `*` in separate fields for easier accesibility.

**How does it work?**
When we create a new Authorization we should check if we already have Authorizations for this ownerKey + resourceType + permissionType stored in the state. If we don't we create a new one. 
If we do we fetch it, check if the permission doesn't exist yet and if not add it to the Authorization object before updating it in the state.

Deleting an Authorization will be similar. We fetch the Authorization from the state by ownerKey + resourceType + permissionType. If it doesn't exist we can reject the command. If it does exist we can retrieve the permissions from it. If the permission to delete doesn't exist we reject the command. If it does, we remove it and update the Authorizations.

--------------

This does require some extra complexity on Authorization creation and deletion, but it will reduce complexity and increase performance on permission validation. In my opinion that's a good tradeoff.

Note that I did not discuss this datamodel within the team yet, as no kickoff has been held at this time.

Reply:
Regarding this, we could also fix it in a followup PR